File: app/src/main/java/com/testproject/kaera/ringtestapp/service/command/GetTopSubredditCommand.java
package com.testproject.kaera.ringtestapp.service.command;

import com.testproject.kaera.ringtestapp.RingApplication;
import com.testproject.kaera.ringtestapp.enteties.APIRedditItem;
import com.testproject.kaera.ringtestapp.service.api.GetTopSubredditAction;
import com.testproject.kaera.ringtestapp.service.cache.StaticCache;

import java.util.List;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class GetTopSubredditCommand extends Command<List<APIRedditItem>> {

    @Inject Janet janet;
    @Inject StaticCache cache;

    private int count;
    private String after;
    private List<APIRedditItem> receivedData;

    public GetTopSubredditCommand() {
        this(null, 0);
    }

    public GetTopSubredditCommand(String afterId, int count) {
        RingApplication.getComponent().inject(this);
        this.count = count;
        this.after = afterId;
        this.receivedData = cache.readData();
    }

    @Override protected void run(CommandCallback callback) throws Throwable {
        if (!receivedData.isEmpty()) {
            callback.onProgress(50);
        }

        janet.createPipe(GetTopSubredditAction.class)
                .createObservableResult(new GetTopSubredditAction(count, after))
                .map(action -> action.getResult())
                .doOnNext(data -> receivedData.addAll(data))
                .doOnNext(data -> cache.writeData(receivedData))
                .subscribe(callback::onSuccess, Throwable::printStackTrace);
    }

    public List<APIRedditItem> getCachedData() {
        return receivedData;
    }
}

File: app/src/main/java/com/testproject/kaera/ringtestapp/service/command/GetTopSubredditCommand.kt
package com.testproject.kaera.ringtestapp.service.command

import com.testproject.kaera.ringtestapp.enteties.APIRedditItem
import com.testproject.kaera.ringtestapp.service.api.GetTopSubredditAction
import com.testproject.kaera.ringtestapp.service.cache.StaticCache
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

/**
 * Created by Dmitriy Puzak on 5/8/18.
 */
@CommandAction
class GetTopSubredditCommand constructor(var afterId: String? = null, var count: Int = 0, var fromCache: Boolean = false) : Command<List<APIRedditItem>>() {

    @Inject
    lateinit var janet: Janet
    @Inject
    lateinit var cache: StaticCache

    private var receivedData: List<APIRedditItem>? = null

    override fun run(callback: CommandCallback<List<APIRedditItem>>) {
        if (fromCache && receivedData != null) {
            callback.onSuccess(receivedData)
            return
        }

        janet.createPipe(GetTopSubredditAction::class.java)
                .createObservableResult(GetTopSubredditAction(count, afterId))
                .map { it.result }
                .flatMap { cache.writeData(it) }
                .doOnNext { receivedData = it }
                .subscribe(callback::onSuccess, Throwable::printStackTrace)
    }

}